x/dymns/keeper/hooks.go

```diff
@@ -0,0 +1,370 @@
+package keeper
+
+import (
+	"errors"
+
+	"github.com/cometbft/cometbft/libs/log"
+
+	errorsmod "cosmossdk.io/errors"
+
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	dymnstypes "github.com/dymensionxyz/dymension/v3/x/dymns/types"
+	rollapptypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	epochstypes "github.com/osmosis-labs/osmosis/v15/x/epochs/types"
+)
+
+/* -------------------------------------------------------------------------- */
+/*                              x/epochs hooks                                */
+/* -------------------------------------------------------------------------- */
+
+var _ epochstypes.EpochHooks = epochHooks{}
+
+type epochHooks struct {
+	Keeper
+}
+
+// GetEpochHooks returns the epoch hooks for the module.
+func (k Keeper) GetEpochHooks() epochstypes.EpochHooks {
+	return epochHooks{
+		Keeper: k,
+	}
+}
+
+// BeforeEpochStart is the epoch start hook.
+func (e epochHooks) BeforeEpochStart(_ sdk.Context, _ string, _ int64) error {
+	return nil
+}
+
+// AfterEpochEnd is the epoch end hook.
+func (e epochHooks) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	miscParams := e.MiscParams(ctx)
+
+	if epochIdentifier != miscParams.EndEpochHookIdentifier {
+		return nil
+	}
+
+	logger := e.Logger(ctx).With("hook", "After-Epoch-End", "epoch-number", epochNumber, "epoch-identifier", epochIdentifier)
+
+	if miscParams.EnableTradingName {
+		if err := e.processActiveDymNameSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	if miscParams.EnableTradingAlias {
+		if err := e.processActiveAliasSellOrders(ctx, logger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// processActiveDymNameSellOrders process the finished Dym-Name Sell-Orders.
+// Sell-Order will be deleted. If the Sell-Order has a winner, the Dym-Name ownership will be transferred.
+func (e epochHooks) processActiveDymNameSellOrders(ctx sdk.Context, logger log.Logger) error {
+	finishedSOs, aSoe := e.getFinishedSellOrders(ctx, dymnstypes.TypeName, logger)
+
+	if len(finishedSOs) < 1 || len(aSoe.Records) < 1 {
+		return nil
+	}
+
+	logger.Info("processing finished SOs", "count", len(finishedSOs))
+
+	for _, so := range finishedSOs {
+		// each order should be processed in a branched context, if error, discard the state change
+		// and process next order, to prevent chain reaction when an individual order failed to process
+		errApplyStateChange := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			if so.HighestBid == nil {
+				e.DeleteSellOrder(ctx, so.AssetId, dymnstypes.TypeName)
+				return nil
+			}
+
+			if err := e.CompleteDymNameSellOrder(ctx, so.AssetId); err != nil {
+				logger.Error("failed to complete sell order", "asset-id", so.AssetId, "error", err)
+				return err
```

Comment:
it's very rare that you should log and return error
you should do one or the other, not both, in 99% of cases

Reply:
I will send you a book on slack

Reply:
In case of production environment, when it very hard to run a node in your machine, as the data size is too large, to debugging the issue, how would you find the issue fast and less effort? So that why I do this.

Can you suggest a way to improve this? `errorsmod.Wrap` something?

Reply:
done